src/image.rs

```diff
@@ -0,0 +1,286 @@
+use hex::encode;
+use hmac::{Hmac, Mac, NewMac};
+use rand::{thread_rng, Rng};
+use sha2::Sha256;
+use std::fmt;
+use std::{collections::HashMap, env::var, thread::sleep, time::Duration};
+use testcontainers::core::{Port, Container, Docker, Image, WaitForMessage};
+
+#[derive(Debug)]
+pub struct ElementsCore {
+    tag: String,
+    arguments: ElementsCoreImageArgs,
+    ports: Option<Vec<Port>>,
+}
+
+impl ElementsCore {
+    pub fn auth(&self) -> &RpcAuth {
+        &self.arguments.rpc_auth
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Network {
+    Mainnet,
+    Testnet,
+    Regtest,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum AddressType {
+    Legacy,
+    P2shSegwit,
+    Bech32,
+}
+
+impl fmt::Display for AddressType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            AddressType::Legacy => "legacy",
+            AddressType::P2shSegwit => "p2sh-segwit",
+            AddressType::Bech32 => "bech32",
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RpcAuth {
+    pub username: String,
+    pub password: String,
+    pub salt: String,
+}
+
+impl RpcAuth {
+    pub fn username(&self) -> &str {
+        &self.username
+    }
+
+    pub fn password(&self) -> &str {
+        &self.password
+    }
+
+    pub fn new(username: String) -> Self {
+        let salt = Self::generate_salt();
+        let password = Self::generate_password();
+
+        RpcAuth {
+            username,
+            password,
+            salt,
+        }
+    }
+
+    fn generate_salt() -> String {
+        let mut buffer = [0u8; 16];
+        thread_rng().fill(&mut buffer[..]);
+        encode(buffer)
+    }
+
+    fn generate_password() -> String {
+        let mut buffer = [0u8; 32];
+        thread_rng().fill(&mut buffer[..]);
+
+        encode(buffer)
+    }
+
+    fn encode_password(&self) -> String {
+        let mut mac = Hmac::<Sha256>::new_varkey(self.salt.as_bytes()).unwrap();
+        mac.update(self.password.as_bytes().as_ref());
+
+        let result = mac.finalize().into_bytes();
+
+        encode(result)
+    }
+
+    pub fn encode(&self) -> String {
+        format!("{}:{}${}", self.username, self.salt, self.encode_password())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ElementsCoreImageArgs {
+    pub server: bool,
+    pub network: Network,
+    pub print_to_console: bool,
+    pub tx_index: bool,
+    pub rpc_bind: String,
+    pub rpc_allowip: String,
+    pub rpc_auth: RpcAuth,
+    pub accept_non_std_txn: Option<bool>,
+    pub rest: bool,
+    pub validatepegin: bool,
+}
+
+impl Default for ElementsCoreImageArgs {
+    fn default() -> Self {
+        ElementsCoreImageArgs {
+            server: true,
+            network: Network::Regtest,
```

Comment:
The networks are named differently for elements, no?

Reply:
True, slightly different and regtest can be an arbitrary string.